<html xmlns:concordion="http://www.concordion.org/2007/concordion">
<head>
    <link href="../../../../concordion.css" rel="stylesheet" type="text/css" />
    <title>"Fail-fast" option</title>
</head>
<body>

    <h1>"Fail-fast" with specific exceptions</h1>

    <p> Fixtures can be annotated with <code>@FailFast</code> to stop further
        processing if any exception is encountered. (<a href="FailFast.html">More details</a>) 
    </p>

    <p> However, it can be useful to fail-fast on some exceptions but not others. 
        The <code>@FailFast</code> annotation can be supplied with a list of 
        exception types. In this case, processing will only be stopped if an exception 
        is encountered that is an instance of one of the exception types in the list.
    </p>
    
    <div class="example">

        <h3>Example - non matching exception</h3>

        <p>When the fixture is annotated with <code>@FailFast(onExceptionType={RuntimeException.class, IOException.class})</code> and processes the following:</p>

<pre class="html" concordion:set="#fragment">
&lt;p concordion:execute="myTimeoutExceptionThrower()"&gt;
This method throws a TimeoutException (which is not a subclass of RuntimeException or IOException).
&lt;/p&gt;

&lt;p concordion:execute="myMethod()"&gt;
This method executes quietly.
&lt;/p&gt;
</pre>

        <p>
            It <b concordion:assertEquals="myMethodWasCalledProcessing(#fragment) ? 'calls' : 'does not call'">calls</b>
            <code>myMethod()</code> in the Java fixture code because the <code>TimeoutException</code> thrown by the
            first method call <b>does not</b> stop further processing of the test.
        </p>

    </div>
    
    <div class="example">

        <h3>Example - matching exception</h3>

        <p>When the fixture is annotated with <code>@FailFast(onExceptionType={RuntimeException.class, IOException.class})</code> and processes the following:</p>

<pre class="html" concordion:set="#fragment">
&lt;p concordion:execute="myFileNotFoundExceptionThrower()"&gt;
This method throws a FileNotFoundException (which is a subclass of IOException).
&lt;/p&gt;

&lt;p concordion:execute="myMethod()"&gt;
This method executes quietly.
&lt;/p&gt;
</pre>

        <p>
            It <b concordion:assertEquals="myMethodWasCalledProcessing(#fragment) ? 'calls' : 'does not call'">does not call</b>
            <code>myMethod()</code> in the Java fixture code because the <code>FileNotFoundException</code> thrown by the
            first method call <b>stops</b> further processing of the test.
        </p>

    </div>


</body>
</html>